pandas/io/packers.py

```diff
@@ -203,18 +202,21 @@ def read(fh):
             with open(path_or_buf, 'rb') as fh:
                 return read(fh)
 
-    # treat as a binary-like
     if isinstance(path_or_buf, compat.binary_type):
+        # treat as a binary-like
         fh = None
         try:
-            fh = compat.BytesIO(path_or_buf)
-            return read(fh)
+            # We can't distinguish between a path and a buffer of bytes in
+            # Python 2 so instead assume the first byte of a valid path is
+            # less than 0x80.
+            if compat.PY3 or ord(path_or_buf[0]) >= 0x80:
```

Comment:
> this would fail for a 0-len buffer. what does the 0x80 compare against? is this platform dependent?

According to the [msgpack spec](https://github.com/msgpack/msgpack/blob/master/spec.md) "Applications can assign 0 to 127 to store application-specific type information.". I believe pandas doesn't currently use this so this assumes if the first byte is below `0x80` it was supposed to be a filename rather than a collection of bytes to decode.